File: Python/271__Encode_and_Decode_Strings.py
class Codec:

    def encode(self, strs):
        """Encodes a list of strings to a single string.

        :type strs: List[str]
        :rtype: str
        """
        if len(strs) == 0:
            return ""
        result = ""
        istr = "_"
        for s in strs:
            istr += str(len(result)) + "."
            result += s
        istr = istr[:len(istr) - 1]
        result += istr
        return result


    def decode(self, s):
        """Decodes a single string to a list of strings.

        :type s: str
        :rtype: List[str]
        """
        if s == "":
            return []
        ind = 0
        for i in range(len(s) - 1, -1, -1):
            if s[i] == "_":
                ind = i
                break

        istr = s[ind:]
        estr = s[:ind]

        inds = istr.lstrip("_").split(".")
        dstr = []
        prev = 0
        for i in range(len(inds)):
            start_ind = int(inds[i])
            if i == len(inds) - 1:
                end_ind = len(estr)
            else: end_ind = int(inds[i + 1])
            t = estr[start_ind : end_ind]
            dstr.append(t)
        return dstr

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(strs))

File: Java/271__Encode_and_Decode_Strings.java
public class Codec {

    // Encodes a list of strings to a single string.
    public String encode(List<String> strs) {
        String result = "";
        for (String str : strs) {
            result += Integer.toString(str.length()) + "|" + str;
        }
        return result;
    }

    // Decodes a single string to a list of strings.
    public List<String> decode(String s) {
        List<String> result = new ArrayList<>();
        int left = 0, right = 0;
        while (right < s.length()) {
            if (s.charAt(right) == '|') {
                int len = Integer.valueOf(s.substring(left, right));
                result.add(s.substring(right + 1, right + len + 1));
                left = right + len + 1;
                right = left;
            }
            right++;
        }
        return result;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(strs));
